### PR TITLE
Added blogpost and sample view for cron runtime monitoring

### PR DESCRIPTION
**Pull request by gagan93jtg**

### CODE REVIEW

**Review by beeeku — APPROVED** [review]

**Review by beeeku — APPROVED** [review]

File: _posts/2019-05-03-monitoring-cron-runtime-in-rails.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+---
+layout: post
+title: Monitoring Cron Runtime In Ruby
+categories: [ruby, whenever, redis, rake, active_admin]
```
Comment (ShubhamGupta) [inline _posts/2019-05-03-monitoring-cron-runtime-in-rails.md]: Lets keep categories very focussed. For example: This is not really a post on Redis.

File: _posts/2019-05-03-monitoring-cron-runtime-in-rails.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+author: gagan93
+---
+
+We have our web backend in Ruby on rails, and we use many gems for different purposes. One such useful gem is **[whenever](https://github.com/javan/whenever)**. We have hundreds of cronjobs running per day. As our system scaled, it was essential to make sure that cronjobs run in the expected time. For, e.g., if a cronjob has to run every hour, we must make sure that one copy must run in less than an hour so that there are no cases when multiple copies of the same cron are running. Though whenever provided a very clean syntax for defining and deploying cronjobs, there was no insight on what’s going on. So we started building a simple in-house solution.
```
Comment (ShubhamGupta) [inline _posts/2019-05-03-monitoring-cron-runtime-in-rails.md]: 1. Lets make hundreds of cronjobs to "thousands".
2. Example can be more clear. One use case is -- There should not be more than one occurrence running at a time. Other case is that if expected runtime is 5 min, then we should get notified if it exceeds. This second thing is not discussed.
3. We should add a one-liner for what whenever does. Something like "We use whenever to manage these cron tasks."

File: _posts/2019-05-03-monitoring-cron-runtime-in-rails.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+
+## Calculating cron runtime
+[whenever](https://github.com/javan/whenever) provides two ways of running rails scripts: rake tasks and runners. One easy way to calculate runtimes is by using rake hooks. Rake provides start and end hooks which we can use to check the start time and end time of a cron job. So first, we moved all whenever crons to rake based scripts (we had few jobs which were written using runner syntax). However, these hooks run on all rake tasks (some of which may not be cron jobs), so we namespaced cron jobs properly to make this hook run for cron based rake tasks only. Also, we used Redis hashes to store start_time, end_time and task details. Here is the sample code,
+
+```ruby
+# Rakefile
+tasks = Rake.application.tasks
+tasks.each do |task|
+  next unless task.to_s.starts_with?('loconav_crons:')
```
Comment (ShubhamGupta) [inline _posts/2019-05-03-monitoring-cron-runtime-in-rails.md]: Lets not use `loconav` here.

File: _posts/2019-05-03-monitoring-cron-runtime-in-rails.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+```ruby
+require './app/modules/cron/execution_constraints.rb'
+
+::Cron::ExecutionConstraints::SCHEDULES.each do |schedule, constraints|
+  options = {}
+
+  if constraints[:run_at_ist]
+    constraints[:run_at_ist] = [constraints[:run_at_ist]] unless constraints[:run_at_ist].is_a? Array
+    run_at_utc = constraints[:run_at_ist].map{ |run_at_ist| (Time.parse(run_at_ist) - 19800).strftime('%H:%M %P') }
```
Comment (ShubhamGupta) [inline _posts/2019-05-03-monitoring-cron-runtime-in-rails.md]: Remove IST related info.

**Review by ShubhamGupta — CHANGES_REQUESTED** [review]

File: _posts/2019-05-03-monitoring-cron-runtime-in-rails.md (hunk 2 above)
Comment (gagan93jtg) [reply _posts/2019-05-03-monitoring-cron-runtime-in-rails.md]: 1. Sure
2. Mentioned multiple copies case also. Whenever supports that already.
3. It is explained in first para, a little later

> Though whenever provided a very clean syntax for defining and deploying cronjobs, there was no insight on what’s going on

>

**Pull request merged**